### ISSUE TITLE
Markdown parser doesn't recognize inline code `x | y` inside table cell

### ISSUE THREAD

**Issue by ljw1004** [tpetricek/FSharp.Formatting#388]
The Markdown parser doesn't parse this table correctly:

| alpha | beta |
|-------|------|
| `x + y` | plus |
| `x | y` | or |

I think it should be rendered as a table with three rows, the bottom two rows having inline-code as their first cells. This indeed is how GitHub renders it. It's also how the CommonMark demonstrator renders it at http://commonmark.unicorn.fail/ . (However, I couldn't find any mention of tables in the official CommonMark spec, and the official try.commonmark.com doesn't render tables at all).

But the FSharp markdown parser parses it very differently. It parses it as a table with just TWO rows, and the final row gets turned into a PARAGRAPH.

**Comment by tomcl** [comment]
So I have this problem too. Looking at the code, the problem is that table column delimiters are parsed before the inline math parsing.

I will certainly (if no-one else deals with this) make a private version which solves the problem in a minimal way by adding a "skip code" layer to the the table row parsing. Doing this properly requires great care because inline we have code, math (several ways) etc and code ticks inside math etc should not be recognised. So give this a week before I'm ready to push.

I'm also writing some enhancement code for this package adding a few pandoc-like features. This is arranged as an entirely separate preprocessor since I reckoned that would be less messy than hacking this code. If there is demand I can release it as an optional switched component in the markdown parser but it will be several weeks before it is in a decent state.

**Comment by ljw1004** [comment]
I kludged around the problem by preprocessing my markdown, looking for every line that starts with a pipe symbol, replacing it with "ceci_n'est_pas_une_pipe" :), running it through FSharp.Formatting, then replacing it again...

**Comment by tomcl** [comment]
OK - so I have a nice unkludged patch to fsharp.markdown/MarkdownParser.fs that deals comprehensively with the problem. Basically the code to parse a table row was previously a Split on delimiter - I've replaced this by a recursive parser using the already defined recognisers for inline code and math so that delimiters inside other inline stuff are skipped. About 35 lines extra + one change to the existing table row parser

It passes all existing tests + a new kludged test with some complex tables.

Unfortunately I don't really understand github so am not going to try to pull this back to the main github project without a bit of checking that my development process is proper:
(1) I cloned and then forked the main repo which I think was a mistake - I should have forked and then cloned?
(2) having got a cloned fork (or, I guess should be forked clone) on local windows disk I then  run the top-level windows bat file build multiple times. Until it builds all clean.
(3) I then develop the patch using the given VS .sln (which seems to work)
(4) one reason for my unhappiness with this is that after a clean 50% of time when I build this solution after a clean, targets fail with "unexpected error creating debug data .pdb" message. However multiple build will mend this till all 20 projects are  correctly built. After a source code change in fsharp.markdown/markdownparser.fs a build will sometimes work and sometimes fail. Again repeated rebuilds will always work.
(5) the patched code will (after multiple runs) correctly build using the top-level build file.

(6) PS I'm using F# 4.0

(7) I'll try again, forking and then cloning the repo, adding file, rerunning tests, pulling, if told that is right thing to do. I'm sure i'll get the hang of this soon...

Anyway, travails with my VS setup aside, which don't affect code quality - the patched file is attached.
[MarkdownParser.zip](https://github.com/tpetricek/FSharp.Formatting/files/201260/MarkdownParser.zip)


Best wishes, Tom

**Comment by matthid** [comment]
@tomcl  Would love to see that patch integrated. Usually the order doesn't matter that much in git, you can add your remote afterwards with `git remote add <myremotename> url` and push later with `git push <myremotename> <branch>`.

About the "unexpected error creating debug data .pdb" I have never seen it ...
Usually you can call the bat file (or `paket restore`) and then open the solution to develop & run tests. Then you commit locally and push to your fork. After that you can create a pull request on github. Which then will be reviewed and merged. If you have technical problems you can ask on the gitter chat as well https://gitter.im/tpetricek/FSharp.Formatting.

BTW: Github has a lot of very good resources to learn git: https://help.github.com/articles/good-resources-for-learning-git-and-github/. They even have one incredibly interactive guide in there (https://try.github.io/levels/1/challenges/1),

**Comment by tomcl** [comment]
Patch pull request done...

**Comment by tomcl** [comment]
OK - so my pull request should close this issue - though it does not deal with the orthogonal star in code block in list issue

Re .pdb error. This is windows 10/VS community 2015. It seems that FAKE cannot use compiler services reliably when VS2015 is running (even with no solution open) but does when VS2015 is not running. The build script runs fine as long as VS is not running.

Re adding tests for my fix. I've added one test to tests/fsharp.markdown.tests/markdown.fs which seems to be the right place to do this - but I'm not entirely sure that the tests there are properly integrated since running build I can't tell whether they pass or fail, unlike the metaprogramming tests.

Re the star in code block in list error. I have not looked at this much. It is more complex because the parse tree spec for lists would have to change to allow nested code blocks as well as nested lists. It is in a different part of the code so I'd need to work out exactly how the existing list nesting happens before touching it.

I've now tidied up by deleting the extra directory under tests.
